### ISSUE TITLE
gRPC Transport

### ISSUE THREAD

**Issue by sharpninja** [CoreWCF/CoreWCF#198]
I'm about to embark on creating a gRPC transport.  Any advice before I start?

**Comment by FreeAndNil** [comment]
Using WCF or CoreWCF seems not to be the right decision for that. WCF is all about SOAP as protocol. Maybe you can use asp.net core as host.

**Comment by Grauenwolf** [comment]
The whole point of WCF is that it doesn't care what the underlying transport is. That's why WCF supported Named Pipes, MSMQ, and other non-SOAP formats. 

In my opinion, 3rd party protocols such as gRPC are essential for the long term adoption of CoreWCF.

So as long as it's not distracting the core team from their efforts at completing v1.0, experimentation should be encouraged.

**Comment by RyoukoKonpaku** [comment]
If I recall there was also talks here around that idea regarding MSMQ in a way that it would be easy to make another Queue Transport (e.g. RabbitMQ, AzureSB) and you wouldn't have to change anything (at least substantial) on the consumer side. 

+1 for that as well and I'm pretty sure it's in line with what WCF's supposed name implies (a multi-transport communication framework).

**Comment by Grauenwolf** [comment]
Last I heard, Non-HTTP protocols such as MSMQ are low priority. Apparently the internals of CoreWCF are focused on an HTTP pipeline, so some post 1.0 work is needed for before MSMQ can be started.

**Comment by sharpninja** [comment]
Fortunately gRPC is HTTP based and I've spent a lot of time thinking about generating the gRPC metadata from the WCF-centric attributes.  I need to sit down and map the specific features of gRPC and how they are currently implemented in the HTTP project and make a list of things that have to change.  

I'm thinking a Roslyn Analyzer will be best for creating the metadata document and then the sinks will bind to the reference gRPC implementation to expose the service and client endpoints.

**Comment by mconnew** [comment]
GRPC has definitely been on my long term plans. The initial release is focusing on NetTcp and Http transports as that's what's most easily supported on ASP.NET Core 2.1 (which is still supported on .NET Framework so makes migration easier). Beyond the initial release, I will be moving to the latest (presumably ASP.NET Core 5.0). Speaking with one of the main developers for ASP.NET Core, it will be a lot easier to support other non-socket based transports.  
While technically true that WCF can be used for non-SOAP framing protocols, the current approach in WCF has been to shoehorn everything to look like SOAP. This can be seen in the WebHttpBinding binding. I plan to fix this after the initial release is "done". I plan to refactor the Message class and make SOAP be one implementation of many.  

Here's how WebHttpBinding pulls off the trick, and you can use a similar approach for GRPC for an design. There shouldn't be too much reworking once I refactor the Message class so it won't be wasted effort. So a quick overview of the Message class. It has 3 main components. Headers, Body and Properties. Properties is a key/value pair store for extra useful information. Headers is meant to map to the SOAP headers, and Body is supposed to map to the SOAP body.  
For REST (WebHttpBinding), from a WCF perspective you have three parts of the HTTP request.  
1. The first header line containing the HTTP verb, the url and the HTTP version.
2. HTTP headers
3. HTTP entity body

Here's how it gets mapped. An HttpRequestMessageProperty is created which holds a copy of all the headers, the verb, the the request/query url and is added to the Properties of the Message. WebHttpBinding does pattern matching to match the request to the operation, Once the operation is known, the Action for the operation is artificially added to the Message headers as the Action header. The request url is rebuilt (request scheme (http/https), Host header and url minus query parameters are combined and added as the To header. The type of Message object used is the concrete type InternalByteStreamMessage which wraps the http request entity. This implementation does a bit of a trick with the Body. There are two ways you can read a message body. One is by calling Message.GetBody<T>. It only supports GetBody<byte[]> or GetBody<stream>. Either way, you can get the bytes of the http request entity back calling either of those. The other way is by calling Message.GetReaderAtBodyContents(). You get returned an XmlReader which is positioned at the <body> tag of the SOAP message. InternalByteStreamMessage fakes this out by returns an XmlReader which fakes there being an opening and closing body tag around the entity body. So when you read into the contents of the "body" element, you can then call XmlReader.ReadContrentAsBase64\[Async\] and the fake XmlReader will return you the byte[] of the entity body.   

The second half of the trick is in the OperationFormatter. This class takes in a Message object and populates an array which will contain the operation parameters. So there's a custom formatter which pulls out any parameters out of the url, and also handles deserializing the message body if needed. it also does it the other way around, it takes the response object, along with any ref or out parameters and creates a Message object that the transport can deal with.  

So to sum it up, you need a custom MessageEncoder and Transport working together to create a Message object with the needed pieces stored some place in it. Routing the request is done using the To and the Action headers so make sure they are populated. Then an OperationFormatter which can pull those pieces out to populate the request parameters. Then do the same thing in return.

**Comment by xprzemekw** [comment]
BTW: If anyone wants to use ServiceMode attributes to create gRPC client/server there is the https://github.com/protobuf-net/protobuf-net.Grpc project.

**Comment by mconnew** [comment]
In theory you don't need to use the WCF attributes in order to have a client. I did a bit of searching and there is a gRPC client on UWP, so you should be able to use a client doing things the gRPC way and still have the service side using the contracts. As long as you can get the protobuf IDL, you should be able to create a UWP client which talks with it. You just won't have the niceness of the WCF contracts.
